keras_nlp/metrics/perplexity.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Perplexity metric implementation based on `keras.metrics.Metric`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class Perplexity(keras.metrics.Metric):
+    """Perplexity metric.
+
+    This class implements the perplexity metric. In short, this class calculates
+    the cross entropy loss and takes its exponent.
+    Note: This implementation is not suitable for fixed-size windows.
+
+    Args:
+        name: string. Name of the metric instance.
+        dtype: string or tf.dtypes.Dtype. Precision of metric computation. If
+               not specified, it defaults to tf.float32.
+        from_logits: bool. If True, `y_pred` (input to `update_state()`) should
+            be the logits as returned by the model. Otherwise, `y_pred` is a
+            tensor of probabilities.
+        pad_token_id: int. Token ID of the padding token. If provided, the mask
+            is computed by this class (all padding tokens are masked while
+            computing the cross entropy loss). Note that if this field is
+            provided, the `sample_weight` field in `update_state()` is ignored.
```

Comment:
This behavior is problematic; we should *combine* the masks, not drop one of them.

Reply:
What do you think is the best way to combine the masks? Element-wise maximum of element-wise addition? Or do you have something else in mind?

Reply:
I think we can just multiply the masks together. If padding token is set, that will give a mask of 1s and 0s, which could be multiplied with sample_weight.

Put one way... If a padding token has sample weight `0.5` it should be ignore. If a non padding token has sample weight 0.5 we should still weight it by its sample weight before summing it in.

Reply:
Great! Done 👍🏼